### DISCUSSION TITLE
objects.inv - available?

### DISCUSSION THREAD

**Discussion by mmacy** [encode/httpx#3091] [Q&A] [answered]
After digging around a bit for the object inventory file (`objects.inv`) for httpx, I'm throwing in the towel and asking outright:

❓ **Q: Is the `objects.inv` for httpx hosted anywhere that dependent documentation projects can reference?**

I didn't see it in the [gh-pages](https://github.com/encode/httpx/tree/gh-pages) branch, and I'm aware of the open discussion in https://github.com/encode/httpx/discussions/1220, but I haven't yet been able to unearth the inventory if indeed you are generating and propping it somewhere.

If it's around and I'm just missing it somehow, would love to find out where. Thanks!

**Comment by lovelydinosaur** [comment]
We don't generate an `objects.inv` file, no.
There might be some useful/interesting things to talk about here, tho I'd like to get a better idea of context first.
Are you able to show an example of what you're aiming for here?

**Reply by mmacy** [reply]
This would enable references to `httpx` types in the API ref of packages that use `httpx` to automatically link to those types in your API ref.

For example, when I build the docs for [openai.OpenAI](https://mmacy.github.io/openai-python/latest/reference/#src.openai.OpenAI), the `Client` type in the image below would (automatically) be a hyperlink to your API reference for that type (because, as also highlighted in the image, it's an `httpx.Client`).

In my case, I'm using [mkdocstrings](https://mkdocstrings.github.io/python/usage/?h=sphinx+inventory#import), but straight Sphinx projects will typically use [sphinx.ext.intersphinx](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) directly for doing this (that Sphinx doc has a good explanation of `objects.inv`).

It's how [`Mapping`](https://docs.python.org/3/library/typing.html#typing.Mapping), [`str`](https://docs.python.org/3/library/stdtypes.html#str), [`object`](https://docs.python.org/3/library/functions.html#object), and [`bool`](https://docs.python.org/3/library/functions.html#bool) in the following image are all linked to the standard library API ref (because Python makes their [`objects.inv`](https://docs.python.org/3/library/functions.html#bool) available to implementers).

![image](https://github.com/encode/httpx/assets/13394541/8a969fd8-fdc8-452d-8bfb-b01c6a10b148)

Here's where I'd pull in your `objects.inv` if it was available:

https://github.com/mmacy/openai-python/blob/6c8443a6a292f4b6bba7bb9d4bd5061cf43d0ed1/mkdocs.yml#L83-L86

**Reply by Ravencentric** [reply]
Please consider this. [It's very helpful for anyone referencing httpx objects in their API documentation](https://mkdocstrings.github.io/usage/?h=inv#cross-references-to-other-projects-inventories)

**Reply by Ravencentric** [reply]
[Here's another example](https://pynyaa.ravencentric.cc/api-reference/clients/#pynyaa.Nyaa.get) where a method may raise `httpx.HTTPStatusError`. With `objects.inv`, mkdocstrings would be able to hyperlink these to the same in httpx's documentation.

**Reply by ItsDrike** [reply]
I'm in the same boat here, I'd also like to cross-reference some httpx classes from my docs to make it clear what these classes are and how they work to the users of my library, however I can't currently achieve that because of the missing `objects.inv`.

I suppose the reason why you don't generate this file is because you're using [`mkautodoc`](https://github.com/tomchristie/mkautodoc) (instead of the more popular [`mkdocstrings-python`](https://mkdocstrings.github.io/python/)), which doesn't support doing so. There isn't currently any issue opened  about adding `objects.inv` generation upstream there, however, from the looks of it, the project doesn't appear to be maintained anymore, with the last commit being 2 years old, so I'm not sure if it's even worth opening this issue there.

Considering the state ot `mkautodocs`, would you at all be willing to consider switching to `mkdocstrings-python`?

**Reply by lovelydinosaur** [reply]
> from the looks of it, the project doesn't appear to be maintained anymore, with the last commit being 2 years old, so I'm not sure if it's even worth opening this issue there.

It's not maintained. Needs an issue/PR prompting it to move into properly deprecated and pointing at more active alternatives.

> Considering the state ot mkautodocs, would you at all be willing to consider switching to mkdocstrings-python?

Yep.

> I can't currently achieve that because of the missing objects.inv.

Im somewhat allergic to `objects.inv` because hyperlinking on the web shouldn't need solutions that are language specific.

**Reply by Ravencentric** [reply]
> Im somewhat allergic to objects.inv because hyperlinking on the web shouldn't need solutions that are language specific.

It's a defacto standard in the python community and unfortunately there's no alternative option here other than just not having httpx references hyperlinked in auto generated API docs

**Reply by ItsDrike** [reply]
> Im somewhat allergic to objects.inv because hyperlinking on the web shouldn't need solutions that are language specific.

Why is that? The alternative solution would have to be hard-coding docs urls for every object manually, which would be super annoying. (Although I do think that it'd be nice if mkdocstrings provided an option to at least do that). I'm not hugely familiar with the format of `objects.inv`, so it may be fair to say that it sucks, but not having it, at least at current time when there really isn't any other alternative (not even hard-coding) is definitely worse.

What's especially annoying is that the auto-generated docs often only show the type annotation, which is just `Client`, and it may be hard to infer what kind of client that is, I could enable showing fully qualified names, but that would do so on every type for that function, which isn't ideal, so I have to explain what that parameter is  in the docstring, with proper reference there, that wouldn't be necessary.

**Reply by lovelydinosaur** [reply, marked as answer]
I'm voicing a design sensibility here, rather than a "let's not do that".

In some slightly alternate imagined space we're interlinking across projects using permalinks & redirects...

* `https://www.python-httpx.org/glossary/BasicAuth`
* `https://www.python-httpx.org/glossary/DigestAuth`
* `https://www.python-httpx.org/glossary/NetRCAuth`

Anyways, yep sure let's have an `objects.inv` (it doesn't suck, it's a useful tool & it's what we've got).

And yes, a switch across to [`mkdocstrings`](https://mkdocstrings.github.io/python/) would make sense.

**Comment by samuelcolvin** [comment]
Chiming in here, a `objects.inv` file for HTTPX would be great please @tomchristie.

**Comment by lovelydinosaur** [comment]
Alrighty... @samuelcolvin helpfully nudged this again via email.

> Please can you add objects.inv to HTTPX so we can link to it from pydantic-ai

That sounds interesting. 🧐

It's much easier for me to get a grip on this if there's a real-life example that we can talk about wrt. tooling, rather than just talking abstract. Would anyone be able to point to a couple of examples of projects with publicly accessible `objects.inv` URLs and show how that helps cross-referencing tools.

*(I realise that this might seem onerous/unneccessary to folks fully invested in sphinx-land, tho we're not using that docs tooling over in this part of the ecosystem and I find [the file format amusingly inscrutable](https://sphobjinv.readthedocs.io/en/stable/syntax.html). Also somewhat musing [on @Ravencentric's point](https://github.com/encode/httpx/discussions/3091#discussioncomment-10981896).)*

**Reply by mmacy** [reply]
It's not just sphinx-land. In fact, since your docs are already in mkdocs-material, you could use [mkdocstrings](https://mkdocstrings.github.io/) for your API ref and it'll build and include the `objects.inv` in your built doc site for you automatically.

**Reply by samuelcolvin** [reply]
Thanks @tomchristie. To be clear, we reference httpx types lots in the pydantic-ai documentation, e.g. [here](https://github.com/pydantic/pydantic-ai/blob/b201dcd6e59bd08ef50f8d5a400d3c1f9191c463/docs/models/anthropic.md?plain=1#L63), I'd love to just change that line to

```
You can customize the `AnthropicProvider` with a custom [`AsyncClient`][httpx.AsyncClient]:
```

So it links to httpx's API documentation, this works with lots of other dependencies, since our API docs are generated by mkdocsstrings:

```yaml
          import:
            - url: https://logfire.pydantic.dev/docs/objects.inv
            - url: https://docs.python.org/3/objects.inv
            - url: https://docs.pydantic.dev/latest/objects.inv
            - url: https://dirty-equals.helpmanual.io/latest/objects.inv
            - url: https://fastapi.tiangolo.com/objects.inv
            - url: https://typing-extensions.readthedocs.io/en/latest/objects.inv
            - url: https://rich.readthedocs.io/en/stable/objects.inv
            # waiting for https://github.com/encode/httpx/discussions/3091#discussioncomment-11205594
```

See [here](https://github.com/pydantic/pydantic-ai/blob/main/mkdocs.yml#216).

TL;DR - if you generate the API documentation with [mkdocstrings](https://mkdocstrings.github.io/) you'll get the `objects.inv` file generated automatically.
